analysis/bulk-deconvolution/scripts/run-epic.R

```diff
@@ -0,0 +1,120 @@
+# This script runs EPIC, using both of its built-in gene signatures sets,
+# for all bulk samples in a given ScPCA project.
+# It exports a TSV of cell type proportions for each sample in the project
+# for both the TRef and BRef reference in EPIC.
+# Note that EPIC does not require a seed.
+
+renv::load()
+library(optparse)
+library(EPIC)
+
+# Parse options --------
+option_list <- list(
+  make_option(
+    "--input_file",
+    type = "character",
+    help = "Input RDS file with a TPM matrix."
+  ),
+  make_option(
+    "--output_file",
+    type = "character",
+    help = "Output TSV file to save EPIC inferences to."
+  )
+)
+opts <- parse_args(OptionParser(option_list = option_list))
+
+# Define functions --------------------
+
+# This is a helper function to format the cell fraction matrix that EPIC infers
+# into a long-format data frame with additional columns for:
+# - the reference (`ref_name`)
+# - the `mRNAProportions` fractions
+# - whether the model for the given sample converged
+format_epic_output <- function(epic_output, ref_name) {
+
+  # First, make a data frame of sample convergence
+  convergence_df <- epic_output |>
+    purrr::pluck("fit.gof") |>
+    tibble::rownames_to_column(var = "sample_id") |>
+    # the convergeCode is like bash: 1 is fail, 0 is success
+    dplyr::mutate(converged = !as.logical(convergeCode)) |>
+    dplyr::select(sample_id, converged)
+  
+  # Extract `cellFractions` into long data frame
+  epic_df <- epic_output |>
+    purrr::pluck("cellFractions") |>
+    as.data.frame() |>
+    tibble::rownames_to_column(var = "sample_id") |>
+    tidyr::pivot_longer(
+      -sample_id,
+      names_to = "epic_celltype",
+      values_to = "fraction"
+    )
+  
+  # Add column with mRNAProportions
+  epic_output |>
+    purrr::pluck("mRNAProportions") |>
+    as.data.frame() |>
+    tibble::rownames_to_column(var = "sample_id") |>
+    # name columns to facilitate joining
+    tidyr::pivot_longer(
+      -sample_id,
+      names_to = "epic_celltype", # this column will be used to join
+      values_to = "mRNAProportions" # this column will actually contain the values
+    ) |>
+    # join back into epic_df
+    dplyr::right_join(epic_df, by = c("sample_id", "epic_celltype")) |>
+    # arrange columns
+    dplyr::select(sample_id, epic_celltype, fraction, mRNAProportions)
```

Comment:
We should have `reference` here too. I think you are missing a join.

Reply:
😬